packages/flutter/test/material/chip_test.dart

```diff
@@ -188,6 +188,91 @@ void _expectCheckmarkColor(Finder finder, Color color) {
   );
 }
 
+Future<void> pumpFrames({
```

Comment:
Why do you need to do this instead of just `tester.pump(duration * frames);`?  As far as I know, it's literally no different.

Reply:
They don't behave the same, but I'm not sure why it happens.

I tried to change line 1086:
    await pumpFrames(tester: tester, frames: 10, duration: const Duration(milliseconds: 20));
To this:
    await tester.pump(const Duration(milliseconds: 200));
But the first choice correctly makes a ripple, while the second choice doesn't. Only the first passes the test.

According to [documentation](https://api.flutter.dev/flutter/flutter_test/WidgetTester/pump.html), `WidgetTester.pump` triggers only one frame after `duration` amount of time, so the first choice simulates 10 frames while the second choice simulates 1 frame.

So it might be that `InkWell` or some other widget used in `Chip` animates the ripple with respect to the **number of frames** passed, instead of the **amount of time passed**. But this is only my guess.

Reply:
No, the animation is definitely based on time:
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/ink_splash.dart#L136
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/ink_splash.dart#L143

It fades out completely in a total of 200ms.

What expectation fails in the test when you say it doesn't make a ripple? It could be that in order to start the ripple, there needs to be a frame pumped, and then another to jump the 200ms.

Can you try this:
```dart
await tester.pump();
await tester.pump(const Duration(milliseconds: 200));
```
and see what happens?

Reply:
Adding another `await tester.pump();` before `await tester.pump(const Duration(milliseconds: 200));` doesn't work.
Also, there is already an `await tester.pump();` before using `pumpFrames`. (Now on line 1080.)

I experimented with different versions of waiting for 200ms; some of them work and some of them *don't* work.

The following versions work:

```
    await pumpFrames(tester: tester, frames: 10, duration: const Duration(milliseconds: 20));
```

```
    await tester.pump(const Duration(milliseconds: 100));
    await tester.pump(const Duration(milliseconds: 100));
```

```
    await tester.pump(const Duration(milliseconds: 50));
    await tester.pump(const Duration(milliseconds: 50));
    await tester.pump(const Duration(milliseconds: 50));
    await tester.pump(const Duration(milliseconds: 50));
```

The correct ripple has center `const Offset(3.0, 3.0)` and radius `3.5`.

The following versions *don't* work:

```
    await tester.pump(); // using one extra tester.pump();
    await tester.pump(const Duration(milliseconds: 200));
```
(Displays a `drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))`)

```
    await tester.pump(const Duration(milliseconds: 60));
    await tester.pump(const Duration(milliseconds: 140));
```
(Displays a `drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))`)

```
    await tester.pump(const Duration(milliseconds: 140));
    await tester.pump(const Duration(milliseconds: 60));
```
(Displays a `drawCircle(Offset(2.6, 2.6), 2.1, Paint(Color(0x66c8c8c8)))`)

```
    await tester.pump(const Duration(milliseconds: 20));
    await tester.pump(const Duration(milliseconds: 180));
```
(Displays a `drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))`)

```
    await tester.pump(const Duration(milliseconds: 20));
    await tester.pump(const Duration(milliseconds: 60));
    await tester.pump(const Duration(milliseconds: 60));
    await tester.pump(const Duration(milliseconds: 60));
```
(Displays a `drawCircle(Offset(2.6, 2.6), 2.1, Paint(Color(0x66c8c8c8)))`)

```
    await tester.pump();
    await tester.pump();
    await tester.pump();
    await tester.pump();
    await tester.pump(); // using 5 extra tester.pump();
    await tester.pump(const Duration(milliseconds: 200));
```
(Displays a `drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))`)

When the test fails, `flutter test` prints out the complete display list like this:
```
          The complete display list was:
            * save()
            * translate(0.0, 0.0)
            * clipRect(Rect.fromLTRB(0.0, 0.0, 370.0, 32.0), clipOp: ClipOp.intersect, doAntiAlias:
true)
            * save()
            * translate(342.0, 4.0)
            * drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))
            * restore()
            * save()
            * clipPath(Instance of 'Path', doAntiAlias: true)
            * drawRect(Rect.fromLTRB(0.0, 0.0, 370.0, 32.0), Paint(Color(0x00bcbcbc)))
            * restore()
            * restore()
            * drawPath(Instance of 'Path', Paint(Color(0x1f000000)))
            * drawParagraph(Instance of 'Paragraph', Offset(342.0, 4.0))
            * drawParagraph(Instance of 'Paragraph', Offset(12.0, 9.0))
            * save()
            * restore()
```

You can modify the test to see this.

Reply:
Thanks for the write up! I'm still don't know _why_ this is the case, but this is interesting.

Reply:
Well, that's very odd.  We should understand what we're actually measuring here before just moving on: it might mean we aren't testing for the right thing.

Can you try one more?

```dart
    await tester.pump(const Duration(milliseconds: 200));
    await tester.pump();
```

Reply:
This one doesn't work:
```
    await tester.pump(const Duration(milliseconds: 200));
    await tester.pump();
```

(Displays a `drawCircle(Offset(2.0, 2.0), 0.0, Paint(Color(0x66c8c8c8)))`)

Reply:
OK, well, for the purposes of this PR, could you please use:
```dart
    await tester.pump(const Duration(milliseconds: 100));
    await tester.pump(const Duration(milliseconds: 100));
```
instead of the `pumpFrames`: it seems less magical and complicated. And could you please file an issue with the above information so we can track down why that is?  I'm suspecting that there is some boundary condition there that we're not taking into account, because I think that the basic 200ms pump should work.

Reply:
@gspencergoog 

Thank you! This works as expected.
I also changed the other `pumpFrames` to `tester.pump` and removed `pumpFrames`. All tests passed.

I'll try to file the issue today or tomorrow.

Reply:
@gspencergoog 

Issue filed — During testing, different ways to wait (tester.pump) the same duration creates different outcomes. #43301